### ISSUE TITLE
kube-proxy cannot be run independently of the kubelet

### ISSUE THREAD

**Issue by smarterclayton** [openshift/origin#6908]
~~The proxy was never upgraded to use the upstream config style approach, which means that we can't even bypass it with extended arguments.~~  We in general need to be able to run the proxy independent (as its own mode) in the future.

In the short term, we should look to do 3 things:
1. Allow the proxy to be disabled via command line flag to openshift start node
2. ~~Allow the Proxy to have a config flag (in either Kube style "extended options" mode or a config in master config) that sets userspace vs iptables~~ (done in #7088)
3. ~~Allow the proxy to take the full set of upstream config flags via extended options, AND~~ have a separate start mode for it `openshift start proxy` that does the same stup as node.

The first would probably be enough to work around to use the old style userspace proxier (we can disable it, then have someone run `openshift start kubernetes proxy` with all the arguments).  The problem there is that the config setup would be hard.  So we probably need to do 2 or 3 above - 3 is probably the best option.  We're going to need to run a new proxier mode anyway for idling (the idler proxy) and so we want to be able to boot it to a new process.

For 3 above we have the problem that multi tenant SDN depends on being loaded with the proxier - it's ok for that config to stay with RunProxy and be flexible.

**Comment by smarterclayton** [comment]
@danwinship for network, @liggitt for setup of the master

**Comment by liggitt** [comment]
right now, I think the SDN plugin and the proxy startup is entangled... would you enable/disable those as a piece, or would we need to run the SDN in one process and the proxy in another?

**Comment by liggitt** [comment]
oh, apparently, I can't read

> For 3 above we have the problem that multi tenant SDN depends on being loaded with the proxier - it's ok for that config to stay with RunProxy and be flexible.

**Comment by smarterclayton** [comment]
I think we'd move SDN to be with proxy for now (--no-proxy on node
doesn't run SDN)

**Comment by dcbw** [comment]
@danwinship I see you just pushed a branch related to this, are you working on it?

**Comment by danwinship** [comment]
Yes. The branch (https://github.com/danwinship/origin/tree/proxy-config) is basically working, I think, other than not actually being configurable:

    if /* FIXME */ true {

but it works with either `true` (iptables) or `false` (userspace) there.

(It doesn't "./tests/extended/networking.sh", but that's because git master doesn't...)

**Comment by abhat** [comment]
@danwinship @liggitt is there an eta on when this will be available as a config option?

**Comment by liggitt** [comment]
https://github.com/danwinship/origin/pull/2 implements 2 (allow toggling between userspace/iptables) and part of 3 (all the options that make sense for our in-process proxy are plumbed). I'll let @danwinship pull it in and make sure it works as he expects. You'd configure it in the node config as an extended arg:
```
proxyArguments:
  proxy-mode:
    - "iptables"
```
or
```
proxyArguments:
  proxy-mode:
    - "userspace"
```

**Comment by abhat** [comment]
@liggitt thanks.

**Comment by liggitt** [comment]
extended arguments for the proxy implemented in #7088

**Comment by liggitt** [comment]
@smarterclayton you want to leave this open to track separating `openshift start node kubelet` and `openshift start node proxy` commands? should we lower priority?

**Comment by smarterclayton** [comment]
Yes and yes

**Comment by smarterclayton** [comment]
Fixed in the previous two issues and the second half of 3 is fixed in #6920
